modin/pandas/groupby.py

```diff
@@ -380,50 +382,33 @@ def _reconstruct_func(func, **kwargs):
                 from pandas.core.base import SpecificationError
 
                 raise SpecificationError("nested renamer is not supported")
-            if isinstance(self._by, type(self._query_compiler)):
-                by = list(self._by.columns)
-            else:
-                by = self._by
-
-            subset_cols = list(func_dict.keys()) + (
-                list(self._by.columns)
-                if isinstance(self._by, type(self._query_compiler))
-                and all(c in self._df.columns for c in self._by.columns)
-                else []
-            )
-            result = type(self._df)(
-                query_compiler=self._df[subset_cols]._query_compiler.groupby_dict_agg(
-                    by=by,
-                    func_dict=func_dict,
-                    groupby_args=self._kwargs,
-                    agg_args=kwargs,
-                    drop=self._drop,
-                )
-            )
-
-            if relabeling_required:
-                result = result.iloc[:, order]
-                result.columns = new_columns
-
-            return result
-
-        if is_list_like(func):
+            func = func_dict
+        elif is_list_like(func):
             return self._default_to_pandas(
                 lambda df, *args, **kwargs: df.aggregate(func, *args, **kwargs),
                 *args,
                 **kwargs,
             )
-        if isinstance(func, str):
-            agg_func = getattr(self, func, None)
+        elif isinstance(func, str):
+            # Using "getattr" here masks possible AttributeError which we throw
+            # in __getattr__, so we should call __getattr__ directly instead.
+            agg_func = self.__getattr__(func)
             if callable(agg_func):
                 return agg_func(*args, **kwargs)
-        return self._apply_agg_function(
-            lambda df, *args, **kwargs: df.aggregate(func, *args, **kwargs),
+
+        result = self._apply_agg_function(
```

Comment:
Is it possible case when `_apply_agg_function` returns `Series` and `relabeling_required` is True? If yes, lines are below `result = result.iloc[:, order]` and `result.columns = new_columns` will throw an exception.

Reply:
Can you provide an example when `DataFrameGroupBy.agg` returns a `Series`? We could add such case to tests.

Reply:
Can you provide an example when `DataFrameGroupBy.agg` returns a `Series`? We could add this case to tests.

Reply:
I can't think of an example right now. However, return value depends on type of member `DataFrameGroupBy._df`, which can be `Series` in some case, but I am not sure regarding this operation. It is not critical for me to merge the PR. We can wait till anyone (users) will be able to face with that case.

Reply:
going by [documentation](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.core.groupby.DataFrameGroupBy.aggregate.html) pandas groupby aggregation **always** returns `DataFrame`, so there is two options:

1. We can believe that every backend returns us DataFrame-like qc (it should do, but that backend responsibility is described nowhere), and then we're okay with the current code
2. If we consider that it's the API layer responsibility to handle Series/DataFrame cases then we can add explicit to frame conversion of the aggregation result

Reply:
Just found out that we're already converting aggregation result to the DataFrame explicitly, so our current code about reordering will workы correct
https://github.com/modin-project/modin/blob/2b0b755753988cefbab38a71db973a26cbe581f4/modin/pandas/groupby.py#L913

Reply:
`self._df` can be either `DataFrame` or `Series` and in depend of that different objects can be created after `type(self._df)(query_compiler=new_manager)`. This is my concern.

Reply:
Although, currently I don't see a case when `self._df` is `DataFrame`, but it was some time ago. At the same time we can get `Series` from `_apply_agg_function` via this [line](https://github.com/modin-project/modin/blob/master/modin/pandas/groupby.py#L917). Since pandas is going to remove `squeeze` parameter in future then we can keep the code without converting to `DataFrame`.

Reply:
Yeah, you're right, I just forgot about that aggregation for Series can also obtain dictionary function...

Then it seems that Modin has a lack of testing Series dictionary aggregation, because yes, it can happen that Series aggregation will return DataFrame:
```python
>>> pandas.Series([1, 2, 3]).groupby(level=0).agg(max=("max"))
   max
0    1
1    2
2    3
```
Then I believe, that because of the line above we will get incorrect result of aggregation (Series instead of DataFrame), however it should be tested.

There is also another problem that we can face, pandas processes dictionary functions differently for 
 [`Series`](https://github.com/pandas-dev/pandas/blob/67a3d4241ab84419856b84fc3ebc9abcbe66c6b3/pandas/core/groupby/generic.py#L231) and [`DataFrame`](https://github.com/pandas-dev/pandas/blob/67a3d4241ab84419856b84fc3ebc9abcbe66c6b3/pandas/core/groupby/generic.py#L940) groupby aggregation, since we use general implementation for DataFrame, this also can cause some problems.

Reply:
So we definitely should add tests for Series groupby dict aggregation, and if them fails and it will be difficult to fix, create an issue and fix that in another PR, since this PR is just about merging dict aggregation with regular one